From 8f78b9efdf7b5e673dd3c089e6c9570c8935f50d Mon Sep 17 00:00:00 2001
From: Lukas Czerner <lczerner@redhat.com>
Date: Thu, 23 Oct 2014 10:12:47 -0400
Subject: [fs] ext4: disable COLLAPSE_RANGE for bigalloc

Message-id: <1414059176-6044-24-git-send-email-lczerner@redhat.com>
Patchwork-id: 98305
O-Subject: [RHEL 7.1 PATCH 23/32] ext4: disable COLLAPSE_RANGE for bigalloc
Bugzilla: 1150171

From: Namjae Jeon <namjae.jeon@samsung.com>

BZ 1150171
https://bugzilla.redhat.com/show_bug.cgi?id=1150171

BREW 8149772
https://brewweb.devel.redhat.com/taskinfo?taskID=8149772

Upstream commit 0a04b248532b358b27a8da050642da6f5f304b03

 Once COLLAPSE RANGE is be disable for ext4 with bigalloc feature till finding
 root-cause of problem. It will be enable with fixing that regression of
 xfstest(generic 075 and 091) again.

 Signed-off-by: Namjae Jeon <namjae.jeon@samsung.com>
 Signed-off-by: Ashish Sangwan <a.sangwan@samsung.com>
 Reviewed-by: Lukas Czerner <lczerner@redhat.com>
 Signed-off-by: "Theodore Ts'o" <tytso@mit.edu>
 Signed-off-by: Lukas Czerner <lczerner@redhat.com>
---

Signed-off-by: Jarod Wilson <jarod@redhat.com>

diff --git a/fs/ext4/extents.c b/fs/ext4/extents.c
index 37c185c..0e73207 100644
--- a/fs/ext4/extents.c
+++ b/fs/ext4/extents.c
@@ -5395,6 +5395,9 @@ int ext4_collapse_range(struct inode *inode, loff_t offset, loff_t len)
  if (!S_ISREG(inode->i_mode))
   return -EINVAL;
 
+ if (EXT4_SB(inode->i_sb)->s_cluster_ratio > 1)
+  return -EOPNOTSUPP;
+
  trace_ext4_collapse_range(inode, offset, len);
 
  punch_start = offset >> EXT4_BLOCK_SIZE_BITS(sb);
-- 
1.7.1